### PR TITLE
Mail Input Output(MIO)

### PR DESCRIPTION
**Pull request by g66shivam**
1. mio.py file contains small mail_client to fetch **unread** emails , received on ps.mio.bits@gmail.com(This mail id can be changed by changing the email id at all instances)
2. mio_index.html is the html template sent to user for acknowledgment mails
3. models.py fill_node_values function has been modified to handle key/value pairs along with HTTPRequest

### CODE REVIEW

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 1]
```diff
@@ -293,77 +293,95 @@ class Node(DjangoDocument):
     use_dot_notation = True
 
 
-    def fill_node_values(self, request=HttpRequest(), **kwargs):
+    def fill_node_values(self, request, **kwargs):
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: Provide some default argument for **`request`** parameter in `fill_node_values` method signature.

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 2]
```diff
@@ -293,77 +293,95 @@ class Node(DjangoDocument):
...
         self.altnames = unicode(altnames)
 
         # 'plural': unicode,
         if kwargs.has_key('plural'):
             plural = kwargs.get('plural', None)
-        else:
+        elif request:
             plural = request.POST.get('plural', None)
+        else:
+            plural = None
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: Give default values based on datatype ('plural': unicode).

> e.g: `plural` is `unicode`
   plural = ''

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 3]
```diff
@@ -293,77 +293,95 @@ class Node(DjangoDocument):
...
 
         # 'member_of': [ObjectId]
         if kwargs.has_key('member_of'):
             member_of = kwargs.get('member_of', [])
-        else:
+        elif request:
             member_of = request.POST.get('member_of', [])
-        self.member_of = [ObjectId(member_of)] if member_of else member_of
+        else:
+            member_of = ''
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `member_of = []`

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 4]
```diff
@@ -378,6 +396,8 @@ def fill_node_values(self, request=HttpRequest(), **kwargs):
                 created_by = kwargs.get('created_by', '')
             elif request:
                 created_by = request.user.id
+            else:
+                created_by = ''
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `created_by = 0`

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 5]
```diff
@@ -388,113 +408,143 @@ def fill_node_values(self, request=HttpRequest(), **kwargs):
                 modified_by = request.user.id
             elif kwargs.has_key('created_by'):
                 modified_by = created_by
-        self.modified_by = int(modified_by) if modified_by else self.created_by
+        else:
+            modified_by = None
+        self.modified_by = int(modified_by) if modified_by else 0
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `modified_by = 0`

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 6]
```diff
@@ -388,113 +408,143 @@ def fill_node_values(self, request=HttpRequest(), **kwargs):
...
 
         # 'contributors': [int]
         if kwargs.has_key('contributors'):
-            contributors = kwargs.get('contributors', [self.created_by])
+            contributors = kwargs.get('contributors', [])
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `contributors = kwargs.get('contributors', [self.created_by])`

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 7]
```diff
@@ -388,113 +408,143 @@ def fill_node_values(self, request=HttpRequest(), **kwargs):
...
+        elif request:
+            contributors = request.POST.get('contributors', [])
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `contributors = request.POST.get('contributors', [self.created_by])`

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 8]
```diff
@@ -388,113 +408,143 @@ def fill_node_values(self, request=HttpRequest(), **kwargs):
...
         else:
-            contributors = request.POST.get('contributors', [self.created_by])
+            contributors = []
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `contributors = [self.created_by]`

File: gnowsys-ndf/gnowsys_ndf/ndf/models.py [hunk 9]
```diff
@@ -388,113 +408,143 @@ def fill_node_values(self, request=HttpRequest(), **kwargs):
...
         self.featured = bool(featured)
 
         # 'url': unicode,
         if kwargs.has_key('url'):
             url = kwargs.get('url', None)
-        else:
+        elif request:
             url = request.POST.get('url', None)
+        else:
+            url = None
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/models.py]: `url = ''`

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+	if name in nodes_list:
+		return error_message["already exists"]
+	else:
+		p.fill_gstystem_values(name = name,member_of=[gst_page._id],created_by = created_by,
+			content = content,group_set=[gst_group._id])
+		p.origin.append({'mio':m_id})
+		try:
+			p.save()
+		except:
+			print "in except"
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: **[Convention]:** You could have better clarity of exception by printing exception itself, in following manner:

<pre>
except Exception as e:
    print "Exception in while creating page from mail:  " + str(e)
</pre>

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+        msg["Message-ID"] = email.utils.make_msgid()
+        msg["In-Reply-To"] = m_id
+
+        if(ref==None):
+                msg["References"] = m_id
+        else:
+                msg["References"] = ref+"\n"+m_id
+
+        msg['Subject'] = subject
+        msg['From'] = "ps.mio.bits@gmail.com"
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: Modify: `msg['From'] = "ps.mio.bits@gmail.com"` with following steps to make it standardize and easy to configure.

1. Add new configuration variable in `settings.py` file: `GSTUDIO_MIO_FROM_EMAIL = ''` and `GSTUDIO_MIO_FROM_EMAIL_PASSWORD = ''`.
2. Let admin to overwrite it in `local_settings.py`.
3. Import `GSTUDIO_MIO_FROM_EMAIL` and `DEFAULT_FROM_EMAIL` from setting.py.
4. Give priority/preference to `GSTUDIO_MIO_FROM_EMAIL` and keep `DEFAULT_FROM_EMAIL` in fallback.
   - If `GSTUDIO_MIO_FROM_EMAIL` is non empty, assign it to `msg['From']`, else use `DEFAULT_FROM_EMAIL`

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+        smtpObj.starttls()
+        smtpObj.login('ps.mio.bits@gmail.com','1Guesswhat')
+        smtpObj.sendmail('ps.mio.bits@gmail.com',to_user,msg.as_string())
+        smtpObj.quit()
+
+def update_page(name,content,id):
+	gst_page = node_collection.one({u'_id':ObjectId(id)})
+	gst_page.name = unicode(name)
+	gst_page.altnames = unicode(name)
+	gst_page.content = unicode(content)
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: Also update/fill following fields:
- `modified_by`
- `modified_at`
- check if current user's id is in `contributors`

OR

Pass arguments to `fill_gsystem_values()`

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+										unique_gs_per_file=True,
+										content_org=content
+										)
+		print "\n\nGS obj id "+str(file_gs_obj._id)+"\n\n"
+		try:
+			# file_gs_obj.save(groupid=home_grp._id)
+			file_gs_obj.save(groupid=home_grp._id)
+		except Exception as e:
+			print 'exception while saving', e
+	else:
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: You want `else` part to do something? If not remove it.

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+		print "file exists"
+	return 'done'
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: **[Convention]:** Possibly, return `Bool (True/False or 1/0)` if you are not returning any data from function. And use same at function calling end.

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+		group_list.append(each.name)
+
+	if(group_name in group_list):
+		gst_home = get_group_name_id(group_name,get_obj=True)
+		if(id in gst_home.author_set):
+			return id,True,error_message["User authenticated"]
+		else:	
+			return id,False,error_message["User not a member of this group"]
+	else:
+		return id,False,error_message["Group does not exist"]
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: You could try following for/replace/in above method:
<pre>
if node_collection.find({ '_type': 'Group', 'name': unicode(group_name), 'author_set': {'$in': [id]} }).count() > 0:
    return id,True,error_message["User authenticated"]
else:
    return id,False,error_message["User not a member of this group"]
</pre>

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+else:
+	shutil.rmtree('attachment')
+	os.mkdir('attachment')
+	print "new made"
+	
+print 'hi'
+def open_connection():
+	try:
+		connection = imaplib.IMAP4_SSL('imap.gmail.com')
+		typ, accountDetails = connection.login('ps.mio.bits@gmail.com', '1Guesswhat')
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: Replace above with variables imported from `settings.py`

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+	def return_update(self):
+		return self.update
+	def return_ObjectId(self):
+		return self.ObjectId
+	def return_fileName(self):
+		return self.Filename
+
+c = open_connection()
+d = open_unseen(c)
+print d
```
Comment (kedar2a) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: **[Convention]:** Either use conveying/explaining variable names or add comments.

File: gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+def authenticate_user(mail,group_name):
+	id = -1
+	try:
+		id = User.objects.get(email=mail).id
+	except:
+		return id,False,error_message["User does not exist"]
+	
+	if node_collection.find({ '_type': 'Group', 'name': unicode(group_name)}).count() > 0:
+		if node_collection.find({ '_type': 'Group', 'name': unicode(group_name), 'author_set': {'$in': [id]} }).count() > 0:
+			return id,True,error_message["User authenticated"]
```
Comment (katkamrachana) [inline gnowsys-ndf/gnowsys_ndf/ndf/views/mio.py]: We may avoid this redundant query by taking the output in any variable.

File: gnowsys-ndf/gnowsys_ndf/settings.py [hunk 19]
```diff
@@ -823,6 +823,12 @@
 # following has to be at last
 # just put every thing above it
 
+#*******MIO_SETTINGS*************
+DEFAULT_MIO_FROM_EMAIL = 'ps.mio.bits@gmail.com'
+GSTUDIO_MIO_FROM_EMAIL = ''
+GSTUDIO_MIO_FROM_EMAIL_PASSWORD = '1Guesswhat'
+#*********************************
+
```
Comment (katkamrachana) [inline gnowsys-ndf/gnowsys_ndf/settings.py]: We have **DEFAULT_FROM_EMAIL** in _settings.py_. That should be used as default mail for MIO too.
What we need will be **GSTUDIO_MIO_FROM_EMAIL** that will be used to override the **DEFAULT_FROM_EMAIL**

**Comment by kedar2a** [comment]
@g66shivam , I have re-tested your commit today. Followings are some findings and suggestions:
- It is working fine for subject with group argument specified :+1: 
- But is is not working if we skip `[group-name]` arg (as we decided it will take Author/User group as fallback group).
  - It's a subject line/text parsing issue. We need to put more logic/flexibility in parsing text/subject-string.
  - e.g: `[Page] sample page`
    - It consider Page (first agr in `[]`) as group name (*code breaking point*).

--

**Strategy we decided:**
- We have thought of having this code in **`alpha`** branch of `gstudio`.
- So I have created fresh branch *mio-alpha* from `alpha` branch and taken pull from your repo's `mio-final` branch.
  - You could find that branch at: [https://github.com/kedar2a/gstudio/tree/mio-alpha](mio-alpha)
  - Also I have done required changes in code in this branch.
  - I would suggest you to get this branch's code in your branch (i.e: `git pull`).

**Comment by g66shivam** [comment]
Hello sir
Nice to hear from you.
I will look into the issue you mentioned and find an appropriate fix for
the same by tomorrow

Thanks

**Comment by kedar2a** [comment]
@g66shivam can you update this pull request with suggested changes?

**Comment by g66shivam** [comment]
@kedar2a 
Sir,I got busy in the college and forgot about the same.
These issues will be resolved by this Sunday.

**Comment by kedar2a** [comment]
:+1:

**Comment by gnowgi** [comment]
data sync is now implemeted through syncthing.  so, this feature is deprecated.

**Pull request closed without merging**